lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -148,70 +142,82 @@ static void createAsyncLoad(scf::ForOp &forOp, tt::LoadOp loadOp, Value alloc,
   }
 }
 
-// Return the transitive use of the load which is a dot operand.
-static Value loadDotOperand(tt::LoadOp loadOp, bool &hasMMAV3) {
-  // We only pipeline loads that have one covert_layout (to dot_op) use
-  // TODO: lift this constraint in the future
-  bool isCandidate = false;
-  if (!loadOp.getResult().hasOneUse())
-    return Value();
-
-  Operation *use = *loadOp.getResult().getUsers().begin();
-  if (auto convertLayout = llvm::dyn_cast<ttg::ConvertLayoutOp>(use)) {
-    auto tensorType =
-        convertLayout.getResult().getType().cast<RankedTensorType>();
-    if (auto sharedEnc =
-            tensorType.getEncoding().dyn_cast<ttg::SharedEncodingAttr>()) {
-      if (sharedEnc.getHasLeadingOffset()) {
-        // MMA V3 case.
-        auto newOrder = sharedEnc.getOrder();
-        auto ty = loadOp.getType().cast<RankedTensorType>();
-        auto oldOrder = ttg::getOrder(ty.getEncoding());
-        if (newOrder[0] == oldOrder[0] || newOrder[1] == oldOrder[1]) {
-          // The operand of MMAv3 is in SharedEncoding and it's order should
-          // not be changed after FuseTranspositions Pass. So we only pipeline
-          // the load if the order of the loaded BlockedEncoding is the same
-          // as the order of the SharedEncoding it is converted to.
-          // TODO: remove this constraint once the LoadOp supports transpose
-          // fusion
-          hasMMAV3 = true;
-          return convertLayout.getResult();
-        }
-      }
+namespace {
+struct LoadDotOperand {
+  LoadDotOperand(tt::LoadOp load,
+                 ttg::DotOperandEncodingAttr dotOperandEncoding)
+      : load(load), dotOperandEncoding(dotOperandEncoding) {}
+  tt::LoadOp load;
+  ttg::DotOperandEncodingAttr dotOperandEncoding;
+};
+} // namespace
+
+// If all the transitive uses of the given value have are used by a convert to
+// the same dot operand encoding, return the encoding. Otherwise return nullptr.
+static ttg::DotOperandEncodingAttr allTransitiveUsesHaveDotEncoding(Value val) {
+  ttg::DotOperandEncodingAttr attr;
+  for (Operation *user : val.getUsers()) {
+    if (user->getNumResults() != 1)
+      return nullptr;
+    auto tensorType = user->getResult(0).getType().dyn_cast<RankedTensorType>();
```

Comment:
Should you have a check if this cast fails ?

Reply:
indeed! Thanks for catching that